db/dcrpg/upgrades.go

```diff
@@ -242,10 +266,132 @@ func compatVersion1Upgrades(db *sql.DB, current, target DatabaseVersion, _ Block
 }
 
 func removeTableComments(db *sql.DB) {
-	for tableName := range createTableStatements {
+	for _, pair := range createTableStatements {
+		tableName := pair[0]
 		_, err := db.Exec(fmt.Sprintf(`COMMENT ON table %s IS NULL;`, tableName))
 		if err != nil {
 			log.Errorf(`Failed to remove comment on table %s.`, tableName)
 		}
 	}
 }
+
+// This upgrade creates a stats table and adds a winners row to the blocks table
+// necessary to replace information from the sqlite database, which is being
+// dropped. As part of the upgrade, the entire blockchain must be requested and
+// the ticket pool evolved appropriately.
+func upgrade110to120(u *updater) error {
+	// Create the stats table and height index.
+	log.Infof("performing database upgrade 1.1.0 -> 1.2.0")
+	exists, err := TableExists(u.db, "stats")
+	if err != nil {
+		return err
+	}
+	if !exists {
+		_, err = u.db.Exec(internal.CreateStatsTable)
+		if err != nil {
+			return fmt.Errorf("CreateStatsTable: %v", err)
+		}
+		_, err = u.db.Exec(internal.IndexStatsOnHeight)
+		if err != nil {
+			return fmt.Errorf("IndexStatsOnHeight: %v", err)
+		}
+		_, err = u.db.Exec(`ALTER TABLE blocks ADD COLUMN IF NOT EXISTS winners TEXT[];`)
+		if err != nil {
+			return fmt.Errorf("Add winners column error: %v", err)
+		}
+	}
+	// Do everything else under a transaction.
+	dbTx, err := u.db.Begin()
+	if err != nil {
+		return fmt.Errorf("failed to create db transaction: %v", err)
+	}
+	defer func() {
+		if err == nil {
+			dbTx.Commit()
+		} else {
+			dbTx.Rollback()
+		}
+	}()
+	makeErr := func(s string, args ...interface{}) error {
+		err = fmt.Errorf(s, args...)
+		return err
+	}
+	// Start with a height-ordered list of block data.
+	blockRows, err := u.db.Query(`
```

Comment:
Didn't know if it's important to do a `queryContext` here or below. The function has the main context as part of the `updater`, but the queries are not particularly heavy.

Reply:
Nah, I'd actually say we don't want it cancellable since it could destroy the DB if half finished.  Same reason why I din't prefer to use the context in the important write queries.